c++/src/kj/mutex.c++

```diff
@@ -364,6 +373,19 @@ void Mutex::unlock(Exclusivity exclusivity, Waiter* waiterToSkip) {
         }
       }
 
+#ifdef KJ_CONTENTION_WARNING_THRESHOLD
+      {
+        uint oldState = __atomic_load_n(&futex, __ATOMIC_RELAXED);
+        uint readerCount = oldState & SHARED_COUNT_MASK;
+        if (readerCount >= KJ_CONTENTION_WARNING_THRESHOLD) {
+          // Atomic not needed because we're still holding the exclusive lock.
+          printContendedReader = true;
+          KJ_LOG(WARNING, "excessively many readers were waiting on this lock", readerCount,
```

Comment:
Maybe we should still delay this `KJ_LOG` until after releasing the lock? So we're not making the readers wait through yet another syscall...

Reply:
Hmm.... I would imagine the backtrace, string formatting, & memory allocations dominate the cost of the actual write syscall. Even still I think it's OK to have this here since this is a rare code path & when I was debugging this issue I would always end up moving this log statement here (makes it easier to set a breakpoint & look at the stack traces of other threads).

The other reason I want this here is that in the future I want to kick off a contention walk of 1 or more shared mutexes to get the lock location or even the whole stack trace of where the readers are at. Usually that information is sufficient to analyze the problem & fix it (barring some sources of real contention I suspect we do have). I don't have immediate plans to do that extension in the near term as I worried about scope creep for now. I want to land the contention walker in EW & make sure it's working to help disambiguate the SIGINT stuff well before we expand its use-case slightly to handle this issue.

Reply:
Oh yeah, the backtrace is also a pretty expensive thing to do while still holding a lock that we know is highly contended.

All I'm suggesting is delaying the actual logging to happen after the futex has been signaled. Yes, I understand it's not common but it's also not hard to defer the logging...